### ISSUE TITLE
Integrate Azure User Identity for credential-less operation on Azure Batch

### ISSUE THREAD

**Issue by abhi18av** [nextflow-io/nextflow#3314]
## New feature

There [are basically three modes of authentication](https://docs.microsoft.com/en-us/azure/storage/common/storage-use-azcopy-authorize-azure-active-directory) which we can explore 

1. Service Principal
2. User-assigned Identity
3. Managed Identity

The user-assigned identity requires certain metadata to be made available to an Azure Resource such as `client ID, Object ID, or Resource ID ` but beyond this, is effectively the same as Managed Identity (a user-identity _managed_ by the Azure platform).

TLDR; The use of user/managed identity 
- Would only require us to do an `azcopy login --managed` within `.command.run` 
- Requires certain properties to be enabled in the Batch account (such as `Identity = SystemAssigned` and `Storage account authentication = "BatchAccountManagedIdentity"`
- Requires the Managed Identity to have sufficient permissions

Managed/User Identity can be used with [all major services](https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/managed-identities-status) which makes sense in Nextflow (and Tower) context such as 
- Monitoring
- Logging
- Managed Disk
- Azure ARC Kubernetes
- Key vault

## Usage scenario 

This enhancement would pave the way for a completely credential less deployment of Nextflow on Azure Batch.

## Suggest implementation 

The foundation of this enhancement would be the `activeDirectory` scope introduced in https://github.com/nextflow-io/nextflow/pull/3132

**Comment by abhi18av** [comment]
Thanks @abhi18av , to address the use-case you have highlighted I have already implemented a service principal based auth here https://github.com/nextflow-io/nextflow/pull/3132 -  I'd love to know hear your thoughts if this would address the immediate concerns?


The use of User-assigned Identity and Managed Identity would build upon the solution implemented in the indicated PR.

**Comment by adamrtalbot** [comment]
Thanks @abhi18av, I was getting a bit mixed up with layers of Microsoft AAD. 

Could the above ticket allow us to run on Azure Batch without _any_ keys? So we could use `az login` or a Managed Identity to authenticate? Or is this scope internal use within Nextflow?

**Comment by abhi18av** [comment]
> Could the above ticket allow us to run on Azure Batch without any keys? So we could use az login or a Managed Identity to authenticate? Or is this scope internal use within Nextflow?

Exactly, that's the plan 💯 

Also, just to clarify the Managed Identity solution afaict could only be used from a VM which has been deployed with proper configuration and then NF head job should be able to offload the responsibility of authentication to the VM's own Managed Identity.

**Comment by adamrtalbot** [comment]
Sounds perfect!

**Comment by pditommaso** [comment]
What's missing to achive this?

**Comment by pditommaso** [comment]
@adamrtalbot this is challenge for you 😉

**Comment by adamrtalbot** [comment]
@abhi18av did you make any progress on this when you were working on it? If I get a chance I will take another look.

**Comment by abhi18av** [comment]
Hey @adamrtalbot , I assume that what you're aiming for is user/managed identity now right?

**Comment by adamrtalbot** [comment]
Yup.

**Comment by abhi18av** [comment]
I'm a bit out of touch on this, perhaps we could sync sometime on Monday?

**Comment by abhi18av** [comment]
Okay, here are the things @adamrtalbot  and @abhi18av  discussed in the meeting today to address the credential-less operations for `azcopy` (and also `fusionfs`) 

- [ ] Restore access to subscription for testing/implementation on the nf-core Azure account

- [ ] Bump azcopy since the version we use is too old https://github.com/nextflow-io/azcopy-tool/pull/1

- [ ] Update the azcopy file transfer logic in Nextflow to rely upon `managed identity` 

- [ ] Possibly enhance the azcopy scope https://github.com/nextflow-io/nextflow/pull/3090/files

**Comment by adamrtalbot** [comment]
> The biggest problem is the authentication for Batch API.

What's missing from the existing [Entra integration](https://www.nextflow.io/docs/latest/azure.html#microsoft-entra-formerly-active-directory-authentication)?

**Comment by pditommaso** [comment]
As long as you need to share the `servicePrincipalSecret` is useless from my of view. We need a share-nothing solution

**Comment by adamrtalbot** [comment]
Yes, especially when every other Azure service can be authenticated by calling `DefaultAzureCredential()` it's frustrating that Batch is left behind.

But these changes would enable more secure storage access, secrets etc.

**Comment by adamrtalbot** [comment]
Thanks Abhinav, can we use a managed identity to authenticate with a batch service? I know we can with storage, key vault etc. but didn't think it was possible with Batch. Some googling suggests it is possible but you have to write some code to do it: https://stackoverflow.com/questions/76326790/how-to-use-user-managed-identity-to-access-azure-batchclient-programatically

**Comment by abhi18av** [comment]
Yes, it is possible. From what I can recall with the efforts done in https://github.com/nextflow-io/nextflow/pull/3132#issuecomment-1231194575 I was able to use managed identity and user-assigned identity through some operations on the Azure portal.

I decided to focus only on the servicePrincipal at the time since `azcopy` also works with it and the PR was becoming too big.

I'll have to poke around in the Git history/branches but I'm sure that this operation was implemented (or sketched out at least) 🤞
